src/express-secure-handlebars.js

```diff
@@ -20,20 +23,123 @@ function ExpressSecureHandlebars(config) {
+ExpressSecureHandlebars.prototype.getTemplates = function (dirPath, options) {
+    options || (options = {});
+    
+    return this._getDir(dirPath, {cache: options.cache}).map(function (filePath) {
+        return path.join(dirPath, filePath);
+    }).then(function(filePaths) {
+        return this._getTemplates(filePaths, options);
+    }.bind(this));
+};
+*/
+ExpressSecureHandlebars.prototype._getTemplates = function (filePaths, options) {
+    options || (options = {});
+
+    var templates = filePaths.map(function (filePath) {
+        return this.getTemplate(filePath, options);
+    }, this);
+
+    return Promise.all(templates).then(function (templates) {
+        return filePaths.reduce(function (hash, filePath, i) {
+            hash[filePath] = templates[i];
+            return hash;
+        }, {});
+    });
+};
+
 
-/* override ExpressHandlebars.render() to expose filePath as compilerOptions */
+// override render() for on-demand partial preprocessing and (pre-)compilation
 ExpressSecureHandlebars.prototype.render = function (filePath, context, options) {
+    options || (options = {});
+
     // expose filePath as processingFile in compilerOptions for secure-handlebars
-    this.compilerOptions || (this.compilerOptions = {});
     this.compilerOptions.processingFile = filePath;
 
-    return expressHandlebars.prototype.render.call(this, filePath, context, options);
+    // ensure getPartials() is called to fetch raw partials
+    return this._exphbsForRawPartials.getPartials(options).then(function(rawPartialsCache) {
+
+        // in constructor, this.compilerOptions.shbsPartialsCache = this._partialsCache
+        this._partialsCache.raw = rawPartialsCache;
+
+        // this._partialsCache.preprocessed is filled by the underlying (pre-)compile() in getTemplate()
+        // TODO: improve this by keeping the returned (pre-)compiled template
+        return this.getTemplate(filePath, options).catch(function(err){
+            throw err;
+        });
+
+    }.bind(this)).then(function() {
+
+        var partialKeys = Object.keys(this._partialsCache.preprocessed);
+
+        // disable preprocessing partials in partial
+        this.compilerOptions.enablePartialProcessing = false;
+        
+        // Merge render-level and (pre-)compiled pre-processed partials together
+        return Promise.all([
+            this._getTemplates(partialKeys, options),   // (pre-)compile pre-processed partials
+            options.partials                            // collected at renderView
+        ]).then(function (partials) {
+            return assign.apply(null, [{}].concat(partials));
+        }).catch(function(err) {
+            throw err;
+        }).finally(function() {
+            // re-enable partial handling
+            this.compilerOptions.enablePartialProcessing = true;
+        }.bind(this));
+
+    }.bind(this)).then(function(partials) {
+
+        options.partials = partials;
```

Comment:
when the call is being called first time, it is the concat of `this._getTemplates(partialKeys, options)` and `options.partials`, so what happen of the second call? it will continue to concat `this._getTemplates(partialKeys, options)`?